File: frontend/src/api/selection.ts
import { fetchSSE, parseSSEStream } from '../utils/sse';

export interface EditSelectionSSEEvent {
  type: 'chunk' | 'done' | 'error';
  data: string;
}

export interface EditSelectionParams {
  projectId: string;
  fullLatex: string;
  selectedText: string;
  instruction: string;
  selectionStart: number;
  selectionEnd: number;
}

export async function* editSelection(
  params: EditSelectionParams,
  signal?: AbortSignal
): AsyncGenerator<EditSelectionSSEEvent> {
  const response = await fetchSSE(
    `/api/v1/projects/${params.projectId}/edit-selection`,
    {
      full_latex: params.fullLatex,
      selected_text: params.selectedText,
      instruction: params.instruction,
      selection_start: params.selectionStart,
      selection_end: params.selectionEnd,
    },
    signal
  );

  for await (const raw of parseSSEStream(response)) {
    try {
      const parsed = JSON.parse(raw.data);
      if (raw.event === 'done') {
        yield { type: 'done', data: parsed.content || '' };
        return;
      }
      if (raw.event === 'error') {
        yield { type: 'error', data: parsed.error || '' };
        return;
      }
      yield { type: 'chunk', data: parsed.content || '' };
    } catch {
      if (raw.data.trim()) yield { type: 'chunk', data: raw.data };
    }
  }
}

File: backend/app/api/v1/selection.py
import json

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sse_starlette.sse import EventSourceResponse

from app.api.schemas.schemas import EditSelectionRequest
from app.core.llm.chains import edit_selection_stream
from app.core.llm.output_parsers import extract_latex
from app.dependencies import get_db, get_project
from app.models.models import Project

router = APIRouter(tags=["selection"])


@router.post("/projects/{project_id}/edit-selection")
async def edit_selection(
    data: EditSelectionRequest,
    project: Project = Depends(get_project),
    db: AsyncSession = Depends(get_db),
):
    async def event_stream():
        full_response = ""
        try:
            async for chunk in edit_selection_stream(
                full_latex=data.full_latex,
                selected_text=data.selected_text,
                instruction=data.instruction,
            ):
                full_response += chunk
                yield {"event": "chunk", "data": json.dumps({"content": chunk})}

            # Strip markdown code-block wrappers (```latex ... ```) if present
            cleaned = extract_latex(full_response)
            yield {"event": "done", "data": json.dumps({"content": cleaned})}

        except Exception as e:
            yield {"event": "error", "data": json.dumps({"error": str(e)})}

    return EventSourceResponse(event_stream())
